scico/denoiser.py

```diff
@@ -0,0 +1,159 @@
+        raise TypeError(f"BM3D requries real-valued inputs, got {x.dtype}")
+
+    # Support arrays with more than three axes when the additional axes are singletons.
+    x_in_shape = x.shape
+
+    if x.ndim < 2:
+        raise ValueError(
+            "BM3D requires two dimensional (M, N) or three dimensional (M, N, C)"
+            f" inputs; got ndim = {x.ndim}"
+        )
+
+    # This check is also performed inside the BM3D call, but due to the host_callback,
+    # no exception is raised and the program will crash with no traceback.
+    # NOTE: if BM3D is extended to allow for different profiles, the block size must be
+    #       updated; this presumes 'np' profile (bs=8)
+    if np.min(x.shape[:2]) < 8:
+        raise ValueError(
+            f"Two leading dimensions of input cannot be smaller than block size "
+            f"(8); got image size = {x.shape}"
+        )
+
+    if x.ndim > 3:
+        if all(k == 1 for k in x.shape[3:]):
+            x = x.squeeze()
+        else:
+            raise ValueError(
+                "Arrays with more than three axes are only supported when "
+                " the additional axes are singletons"
+            )
+
+    y = hcb.call(lambda args: bm3d_eval(*args).astype(x.dtype), (x, sigma), result_shape=x)
+
+    # undo squeezing, if neccessary
+    y = y.reshape(x_in_shape)
+
+    return y
+
+
+class DnCNN(FlaxMap):
+    """Flax implementation of the DnCNN denoiser.
+
+    A flax implementation of the DnCNN denoiser :cite:`zhang-2017-dncnn`.
+    Note that :class:`.DnCNNNet` represents an untrained form of the
+    generic DnCNN CNN structure, while this class represents a trained
+    form with six or seventeen layers.
+    """
+
+    def __init__(self, variant: str = "6M"):
+        """
+
+        Note that all DnCNN models are trained for single-channel image
+        input. Multi-channel input is supported via independent denoising
+        of each channel.
+
+        Args:
+            variant: Identify the DnCNN model to be used. Options are
+                '6L', '6M' (default), '6H', '17L', '17M', and '17H',
+                where the integer indicates the number of layers in the
+                network, and the postfix indicates the training noise
+                standard deviation: L (low) = 0.06, M (mid) = 0.1,
+                H (high) = 0.2, where the standard deviations are
+                with respect to data in the range [0, 1].
+        """
+        if variant not in ["6L", "6M", "6H", "17L", "17M", "17H"]:
+            raise RuntimeError(f"Invalid value of parameter variant: {variant}")
+        if variant[0] == "6":
+            nlayer = 6
+        else:
+            nlayer = 17
+        model = DnCNNNet(depth=nlayer, channels=1, num_filters=64, dtype=np.float32)
+        variables = load_weights(_flax_data_path("dncnn%s.npz" % variant))
+        super().__init__(model, variables)
+
+    def __call__(self, x: JaxArray) -> JaxArray:
+        r"""Apply DnCNN denoiser.
```

Comment:
@crstngc: Can the code here be re-factored so that this method can make use of `FlaxMap.__call__` rather than completely replacing it? If not, is there any real role for the intermediate `FlaxMap` class?

Reply:
The denoiser call is swapping the channel axis with the batch axis to be able to apply a grayscale trained model also to color images. `FlaxMap` does not do that, it directly applies a Flax model. Are you suggesting to always apply Flax models using axes swap?